clang/lib/CIR/CodeGen/CIRGenExpr.cpp

```diff
@@ -653,9 +653,62 @@ RValue CIRGenFunction::buildLoadOfLValue(LValue LV, SourceLocation Loc) {
         getLoc(Loc), load, LV.getVectorIdx()));
   }
 
+  if (LV.isExtVectorElt()) {
+    return buildLoadOfExtVectorElementLValue(LV);
+  }
+
   llvm_unreachable("NYI");
 }
 
+int64_t CIRGenFunction::getAccessedFieldNo(unsigned int idx,
+                                           const mlir::Attribute elts) {
+  assert(elts.isa<mlir::ArrayAttr>());
+
+  auto elt = elts.cast<mlir::ArrayAttr>()[idx];
+  assert(elt.isa<mlir::IntegerAttr>());
+  return elt.cast<mlir::IntegerAttr>().getInt();
+}
+
+// If this is a reference to a subset of the elements of a vector, create an
+// appropriate shufflevector.
+RValue CIRGenFunction::buildLoadOfExtVectorElementLValue(LValue LV) {
+  mlir::Location loc = LV.getExtVectorPointer().getLoc();
+  mlir::Value Vec = builder.createLoad(loc, LV.getExtVectorAddress());
+
+  // HLSL allows treating scalars as one-element vectors. Converting the scalar
+  // IR value to a vector here allows the rest of codegen to behave as normal.
+  if (getLangOpts().HLSL && !Vec.getType().isa<mlir::cir::VectorType>()) {
+    auto DstTy =
+        mlir::cir::VectorType::get(builder.getContext(), Vec.getType(), 1);
+    mlir::Value Zero = builder.getNullValue(builder.getSInt64Ty(), loc);
+
+    Vec = builder.create<mlir::cir::VecCreateOp>(loc, DstTy,
+                                                 llvm::ArrayRef(Zero));
+  }
+
+  const auto Elts = LV.getExtVectorElts();
+
+  // If the result of the expression is a non-vector type, we must be extracting
+  // a single element.  Just codegen as an extractelement.
+  const auto *ExprVT = LV.getType()->getAs<clang::VectorType>();
+  if (!ExprVT) {
+    int64_t InIdx = getAccessedFieldNo(0, Elts);
```

Comment:
Is this path tested?

Reply:
Was this also tested?

Reply:
Yes, I think it's tested by `int one_elem_load = a.s2;`.